### ISSUE TITLE
Retrieving the exact portion of OSM WayId in Map Matching

### ISSUE THREAD

**Issue by nickaein** [valhalla/valhalla#2083]
Meili with `trace_attributes` returns meta-data for the matched route, including `way_id` which specifies the ID of matched Way in the underlying OSM map.

I am looking for a solution to find out **which portion of a Way** (which nodes of Way) is incorporated in matched route. Something likes this would be great:

```
{
    "edges": [
        {
            "way_id": 123456,
            "begin_node_id": 112233,
            "end_node_id": 445566
        },
...
}
```

where `begin_node_id` and `end_node_id` are OSM Node IDs that specify the portion of Way that is selected in the map matching.

**Comment by dnesbitt61** [comment]
Valhalla does not store the OSM node Ids so it cannot return this information. One could extend the NodeInfo structure to add these and then return them in trace_attributes. I don't believe this will be supported in the main project as we have resisted adding OSM node Ids due to the data size increase this would require.

**Comment by nickaein** [comment]
> Valhalla does not store the OSM node Ids so it cannot return this information. 

To be exact, we don't specifically need the OSM Node Ids. Having any indicator (e.g. *index* of start/end node) that specify which portion of an OSM Way is involved in Map Matching would suffice. It seems `trace_attributes` API doesn't provide any of that though.

> One could extend the NodeInfo structure to add these and then return them in trace_attributes. 

I wonder if the necessary data is already present and can be exposed somehow. For instance, 
[`mjolnir::Edge`](https://github.com/valhalla/valhalla/blob/db3a99f2618c0a513fe566f08aac34444829d9cc/valhalla/mjolnir/node_expander.h#L26) apparently references [`mjolnir::Node`](https://github.com/valhalla/valhalla/blob/db3a99f2618c0a513fe566f08aac34444829d9cc/valhalla/mjolnir/node_expander.h#L134) which in turn has an [`OSMNode`](https://github.com/valhalla/valhalla/blob/a3f84dadc85057fc306dcdd3819b099917759b9b/valhalla/mjolnir/osmnode.h#L20) member. Can they be helpful? Or these data structures are unrelated to the data structures which `Meili` engine works with?

P.S. I was a bit surprised that this was not already exposed through Map Matching API. Providing OSM WayIds is nice, but it's not that much helpful. Are we approaching the problem correctly?

**Comment by dnesbitt61** [comment]
The Edge and Node structures in mjolnir are temporary data structures used while building the Valhalla tiles. This data is not available within the tiled graph that Valhalla uses within meili (map matching) or thor (route path computation). Only data structures in baldr are available (e.g. NodeInfo, DirectedEdge, EdgeInfo) and none of these have information like the indexes into the list of OSM nodes that comprise an OSM way.

**Comment by mwoehlke-kitware** [comment]
It turns out I have the exact same problem. I'm trying to use Valhalla as a route planner for traffic simulation using [SUMO](https://en.wikipedia.org/wiki/Simulation_of_Urban_MObility). I can get the OSM way_id's, but SUMO breaks ways into individual segments (at intersections... not one per *node*, thankfully!), and I need some way of deriving those.

**Comment by mwoehlke-kitware** [comment]
Okay, I can give you an inelegant solution.

- First, note also #2460; you're going to want to convert the trip shape to lat/lon if you don't have that already.
- Next, you're going to need to read the OSM using something other than Valhalla. (I used [readosm](https://www.gaia-gis.it/fossil/readosm/index).) Extract all the nodes into a map `id` → `location`. Extract all the ways into a map of `id` → `node ids`.
- Get the locations along the route from the trip shape by trip edge.
- For each trip edge (which gives you the OSM way id), match each point against the locations of the OSM nodes in the associated OSM way. You *could* do a raw match against *all* OSM nodes if you're willing to implement a spatial search, but knowing the OSM way makes it feasible to just do a brute-force match against all points in that way.

There are two "gotchas" with this approach. First, Valhalla *can and will* start and stop trips in the middle of an OSM way in between two nodes, so you'll need to either be careful about getting duplicates for the first and last leg, or else you might want to throw out the first and last location when doing your match. Second, Valhalla *can and will* elide some OSM nodes, so you'll need to be careful to add the intervening nodes when this happens (if that matters to you).

I've solved the latter issue, as well as determining in which direction an OSM way is being traversed, because my application requires that, but how to do so is more complicated than I want to try to explain in a comment. Hopefully I'll have approval to publish our sources soon, at which point you can go look at that. On that note, beware that there's a difference between knowing the OSM node id, and knowing which node of a way a trip leg represents. This is ways *can and do* (sometimes) use nodes more than once. My code ultimately maps a Valhalla trip to the set of {way, way indices} used.

**Comment by nickaein** [comment]
Thanks Matthew for the thorough explanation! We have moved to other tasks for now, but this will be definitely helpful when we return to MapMatching.

**Comment by mwoehlke-kitware** [comment]
FYI, our code [is now available](https://github.com/Kitware/trailblazer/blob/9c03c51fddb0f9ddbc15b93df7df9ba9b10bffb1/trace.cpp)! The "interesting" stuff starts around L247 (everything before that, with the exception of some definitions, is mostly Valhalla route generation, which I assume you already know how to do). The "finicky bits" that I alluded to previously are in [`Graph::locate`](https://github.com/Kitware/trailblazer/blob/9c03c51fddb0f9ddbc15b93df7df9ba9b10bffb1/graph.cpp#L70-L133). You probably don't need the segment matching stuff.

**Comment by Sleepingwell** [comment]
I also had the same requirement and have implemented this facility in Valhalla as described by @dnesbitt61's in the [comment above](https://github.com/valhalla/valhalla/issues/2083#issuecomment-559864420). Please see the last two commits on [this branch](https://github.com/CVTS/valhalla/commits/add-osm-ids-prototype). OSM ids are included in the output for the nodes along each edge and for end nodes. I'm only starting to come to terms with Valhalla, so any comments or advice on this would be welcome!

I only add OSM ids to 'way nodes' (please excuse the parlance if it is dodgy). Other edges are created elsewhere when building the tiles that I have not addressed. I'm not sure if that would make much difference, but for my current use case I'm matching for commercial vehicles (trucks, busses, taxis, ...) only and don't think I'd see much change by going further.

Noting David's comment above, I would still love to get something like this merged. Is there an acceptable way to achieve this (that could be turned off at build time)? One could do it with macros, but in my browsing of Valhalla's code so far, I've not seen a lot of that kind of thing going on and hence wonder if:

- there are other mechanisms used to achieve this that I am unaware of,
- there has simply been little reason to add compile time options, or
- it is frowned upon,

I realise there is work to be done to ready this for a pull request and would be happy to work further on it if there is a chance of success... **please advise**!

**Comment by kevinkreiser** [comment]
@Sleepingwell yes i believe there is a path for optional support of osm node ids in valhallas tiles. while what you have done is not incorrect, i don't believe we would accept it in its current form. instead, i would ask you to add a few things to make support for these ids optional at runtime (at compile time does not make sense imho). Anyway this is what i would propose:

1. in the config builder add a section to mjolnir which is a boolean specifying whether or not to add osm node ids to the tiles, default to false
2. in the graphtile and in the tilebuilder add a new array of 64bit unsigned integers and remove the id that you added directly to the nodeinfo, upon creating a new node in the graph, also push back its osm id if the above config option is set to true
3. during hierarchy building, be sure to copy the nodes' ids (if present) to those nodes which exist on levels other than level 2 (ie 0 1)
4. figure out if transit will break this model. there may be places in the code (though i dont think so) where we add what i would call "false nodes", ie adding a node (shape point) to the graph that actually isnt an osm node. if this happens (i dont think it does) we need to use `std::numeric_limits<uint64_t>::max()` as the magic invalid osm node id. again i dont think we add shape like this but if we do we must insert this invalid id where they are in the list of node ids.
5. what we are left with is, optionally, a parallel array that matches one to one with the array holding the nodeinfos, from there we can always get a nodes id (or invalid id if they were turned off) by using the same offset as the node in the nodeinfo list to look into the optional node id list.

let me know which bits of that didn't make sense and id be happy to elaborate

**Comment by Sleepingwell** [comment]
Thanks @kevinkreiser for the speedy response. I will try and grok your advice and implement in accordance later this week.

**Comment by Sleepingwell** [comment]
I've started working on this but seem to have been stumped by what looks like a bug... but could be that I'm doing something wrong. I've put together a minimal commit to reproduce it, which you can find [here](https://github.com/CVTS/valhalla/tree/demo-serialization-bug), specifically in [this commit](https://github.com/CVTS/valhalla/commit/a5644dbf4df5a4cd58a79bec588625c2f7c46a9c).

The essence is that:

- I have added an extra data member to `EdgeInfoBuilder`,
- I serialise this to the tile (for all EdgeInfoBuilder`s in the tile,),- when the tile is loaded I initialise a data member of `EdgeInfo` with the data,
- I check to see that the relations between the `EdgeInfo` instances are what I'm expecting, which eventually, they are not (I'm expecting that the field increments by one between each instance, but sometimes I get something that looks like uninitialised data).

I've tried to make the example as simple as possible and *think* this is a bug. Would be lovely if you could take and look and let me know if it is, or what I'm doing wrong.

I've added a demo script for the issue. You after checking out that commit, you should be able to reproduce with:

```bash
mkdir build && cd build && cmake .. && cd ../bugdemo && ./test.sh
```

**Comment by kevinkreiser** [comment]
@Sleepingwell we should not modify edgeinfo for this use-case. the only way valhalla allows you to add to the tiles spec is either to:

1. use some of the spare bits in existing data structures (they are usually marked `_spare`)
2. add a new datastructure and use up one of the existing unused spots in the graphtile header (this is what i was suggesting above but actually maybe we should do something smarter)
3. change the structures more like what you were doing with edgeinfo and live with binary incompatibility (we basically never do this, its only happened 2 times, the first of which was going from the first design to a more stable one and the second was an improvement on size and efficiency (and that was like 3.5 years ago))

i know for a first time contributor this can be quite a lot to get a grasp of. at this point i am wondering how i might best show you the ropes a bit more. maybe we should take each chunk of the above suggestion and break it up into more detailed bits.

1. **mark tiles with a bit that tells us that the tile does have osm node ids present.** first we need a place in the graphtileheader to be able to tell us that we should expect osm node ids in the tile. for this we only need 1 bit (yes or no). so we can take it from this spare: https://github.com/valhalla/valhalla/blob/master/valhalla/baldr/graphtileheader.h#L617. all you have to do is give that a name and make a setter and a getter for it. maybe something like `_node_ids` or so?
2. **store osmid in the tilebuilder**. add a vector of uint64_t to the graphtile builder to store the node ids. then, where your current code has the osm id added to the emplace_back (https://github.com/CVTS/valhalla/commit/bbe1f7a9dc76273e68c360ddd6907cd383ad053c#diff-8eb193c3251becf80bdd36d40abdfe7681301bbfbac59302efcbb70f3a2e1fa9R959) you should remove that and add another line that emplace backs the osmid to the graphtile.node_ids() on the next or prior line for example.
3. **have the tile builder know how to serialize and deserialize this info**. there are a couple of serialization methods both in and out for graphtile builder. you need to modify those to notice that the osmnode ids vector is not empty and serialize them out, or in the case that is deserializing (serializing in) you need to check the header bit that we set in step one, and then read the ids into the vector
4. **update graphtile to point to the right place in memory**. add a `uint64_t* _node_ids` field to the graphtile and iin the case that the header has the bit set, set the pointer in the initialize method
5. **update grpahtile to get access to node ids**. add a fucntion to the graphtile that allows you to get an osm node id given an index or a graphid. if the pointer is not null  you can just use it liek an array to return the requested nodes osmid.

now all of this should work **IF** you have heirarchies disabled because in that case all the nodes we ever create live on level 2. this is not the default however, by default we create other levels and when we do that some of the nodes stay on level 2 but others move to levels 1 or 0 because of that we will have to modify the list of osm ids. i didnt refresh my memory but i think what we do in the hierarchy builder is that we crack open a tile and we copy out the stuff to new tiles as we go. i think what this means is that we just need to add code to also copy the osm node ids when we copy the nodes around.

i know this all sounds like a giant pain in the butt, when you could have just added it to the nodeinfo but sadly that is a breaking change and we would not be able to accept that. using the method described above though makes it both optional and supports what you are looking to do, so despite the pain, i hope you're still game to take a crack at it :smile:

**Comment by Sleepingwell** [comment]
@kevinkreiser Thanks again for getting back so quickly.

If one wants to only add OSM ids for the nodes, then I think what your suggesting is straight forward. However, I want all the OSM ids on the edges, and (I think) this is trickier.

Happy to work on whatever solution works, but I don't think the modifications I'm actually proposing would cause binary compatibility problems (the example above would, but that was just a minimal example to demonstrate what I think is a bug). The field I was thinking about *actually* adding to nodinfo object would only be written (and read) if the tile contains OSM ids, so the data in existing tile would still be readable (see  [src/mjolnir/edgeinfobuilder.cc](https://github.com/CVTS/valhalla/commit/b0510b7b48b8c5d825aa996abcf96958cccacaf0#diff-c8173ae35f0cb6abc871bd34619e995eb9f31c01e12fe54f95e59a8af85d91fb) for where the data is written, the corresponding read is also conditional). In that commit, I believe this is the only modification to a core data structure, apart from [adding two extra fields in `GraphTileHeader`](https://github.com/CVTS/valhalla/commit/b0510b7b48b8c5d825aa996abcf96958cccacaf0#diff-02d2f61e72de41c5a1fb592eaf617f2e3c24c1882ed77791a4e876e9a76fd05e) as you suggested. This might still be a bad idea for reasons I don't understand... but just sayin'.

It is not clear we can use the spare slot `EdgeInfo::EdgeInfoInner::spare0_`, as it appears that might not be zeroed out (as they are in GraphTileHeader, for example)... or at least I can't see where that happens and I don't think with the way instances of this struct are declared ensures zero initialization will occur. Hence, that might cause binary incompatibility. The approach I've taken does not require this, however, as the info required to determine if the tile contains OSM ids is in the `GraphTileHeader`.

Please note that the branch I refer to above needs more work. If you do look at it you will notice *a lot* of changes. Many of these are about ensuring that `GraphTileBuilder` knows whether to include OSM ids or not. Some of these are perhaps unnecessary, but it is not clear to me where brand new tiles are created and where they are based on an existing one (which is read first). The other thing that causes the scatter blast of changes is ensuring that every time a new edge is added, OSM ids (or faux OSM ids) are added. I think there may be neater ways of doing this (like using the spare fields in `EdgeInfo::EdgeInfoInner::spare0_`, if we can), but I'm just learning.

**In any case** is the issue I raise in the comment above a bug or am I missing something? I spent ages trying to figure out what I was doing wrong, but as I note, I think this is a bug... and now I need to know! I'd like to finish the path I'm on so we can see how it looks... I think what I have done is *very* close to what have suggested in both your replies (noting that I am also keeping OSM ids along edges). However, it won't work at present because of the issue I raised in my previous comment.

**Comment by kevinkreiser** [comment]
## design of the feature

@Sleepingwell sorry for the confusion, this is a little hard to follow without an open pr and comments on it for example. im not sure if you are ready to but i feel like starting that process might help me understand your approach a bit more. anyway, good that i now understand correctly that you want all the nodes' ids not just the graph nodes' ids. that is unfortunately a lot larger in terms of data and yes a bit trickier. storing it on the edgeinfo in this case does make more sense organizationally although increases the level of duplication of data as well (every graph nodes id will be duplicated for every pair of edges that leave/enter it). anyway it seems like a reasonable idea to start with. lets keep talking about that to see if i can understand how you want to implement it.

one thing that might help would be to just share the proposed modifications to data structures to allow storing and retrieving the ids. from what i've read and seen from the commits it seems you want to conditionally have the edgeinfo object load this data if its present. off the top of my head, i think that could work if we do it the same way we do the shape data. that is, we store it after the edge info, we use the number of shape points to tell how many ids there are, we lazily load by just setting a pointer and letting the caller iterate over it when they want to use it. it seems like maybe what you are thinking too? to get that to work one could:

1. set the bit as i suggested above in the graphtile header to signal the presence of osmids attached to edge info
2. when someone asks for an edgeinfo object from a graphtile, the graphtile sends this boolean to the edgeinfo constructor
3. edgeinfo, getting this flag and also knowing how many shape points it has knows exactly which byte range it should hand back to the caller, we can maybe make use of iteratable_t to set begin pointer and end pointer and give the caller iteration semantics. sadly we cant use `std::span` since we are on c++14
4. when writing nodes osm ids we need to write them after the edgeinfo object they belong to and pad the object to the 4 byte boundary as is currently done to keep the structures byte aligned

yeah i noticed that there were lots of checking of the boolean from the config. im hoping that once we write the tile the first time and the graphtile header has the bit set for having node ids, then we wont need the config item after that because the tile itself will know if the ids are present. but it can wait until you are getting further along with the implementation of course, no worries.

## bug report

as for the bug you've found, ill take a read of it again but looking at the code changes hasn't clicked for me yet what is going wrong there. would you mind splitting this off into another issue so as not to diverge too much from this main issue. we'll get it figured out. i know i did some tricks to support variable length osm way ids perhaps that is messing with the ability to just add stuff onto the end of the structure.

**Comment by Sleepingwell** [comment]
@kevinkreiser OK. Lets take a couple of bites at a time.

I have enough ready for a pull request for simply adding OSM ids for nodes (not for edges). This is reasonably straight forward and I think I'm close to what you have suggested. I've cut the work into more digestible chunks also.

A couple of things remain to do on it, but I'll describe those in the pull request... keep your eyes open.

**Comment by kevinkreiser** [comment]
@Sleepingwell you mean we can do both approaches? one where we let you store just the osm ids for the nodes that become graph nodes and a more detailed mode where we store all the nodes' osm ids that make up all the edges of the graph? i dont see why that wouldnt work. if someone had the config option turned on for both of them i think we'd just do the latter though and not store both sets of information since its duplicated. though i will say dont feel obligated to do both just because i described that other approach, it was just a suggestion before i understood your whole meaning/intent.

anyway i look forward to a PR where we can go over the approach in more detail!
